File: scripts/notion_to_src/parse.py
from __future__ import annotations

import re
from dataclasses import dataclass

_PROPERTY_LINE = re.compile(r"^(.{1,200}?):\s+(.*)\s*$")


def _is_relation_value(val: str) -> bool:
    v = val.strip()
    if "(" in v and (".md" in v or ".csv" in v):
        return True
    if "(/" in v and (".md" in v or ".csv" in v):
        return True
    return False


@dataclass
class SplitPage:
    h1: str
    h1_text: str
    scalar_properties: list[tuple[str, str]]  # raw key, value (lifted to YAML, not in body)
    body_lines: list[str]  # relation property lines + post-property prose, original order per section


def _is_property_line(line: str) -> bool:
    if not line or line.lstrip().startswith("#"):
        return False
    m = _PROPERTY_LINE.match(line)
    if not m:
        return False
    if ":" in m.group(1):
        return False
    return True


def split_notion_page(text: str) -> SplitPage:
    lines = text.replace("\r\n", "\n").split("\n")
    if not lines:
        return SplitPage("# Untitled", "Untitled", [], [])

    h1 = lines[0]
    h1_text = h1
    m = re.match(r"^#\s+(.*)$", h1)
    if m:
        h1_text = m.group(1).strip()

    i = 1
    while i < len(lines) and not lines[i].strip():
        i += 1

    scalars: list[tuple[str, str]] = []
    relation_and_rest: list[str] = []

    if i < len(lines) and _is_property_line(lines[i]):
        while i < len(lines):
            line = lines[i]
            if not line.strip():
                j = i + 1
                while j < len(lines) and not lines[j].strip():
                    j += 1
                if j < len(lines) and _is_property_line(lines[j]):
                    i = j
                    continue
                break
            if not _is_property_line(line):
                break
            m = _PROPERTY_LINE.match(line)
            assert m
            key, val = m.group(1), m.group(2)
            if _is_relation_value(val):
                relation_and_rest.append(line)
            else:
                scalars.append((key, val))
            i += 1

    relation_and_rest.extend(lines[i:])
    return SplitPage(
        h1=h1,
        h1_text=h1_text,
        scalar_properties=scalars,
        body_lines=relation_and_rest,
    )

File: packages/_archive/notion-importer/src/parse.ts
const PROPERTY_LINE = /^(.{1,200}?):\s+(.*)\s*$/;

export interface SplitPage {
  h1: string;
  h1Text: string;
  scalarProperties: [string, string][];
  bodyLines: string[];
}

function isRelationValue(val: string): boolean {
  const v = val.trim();
  if (v.includes("(") && (v.includes(".md") || v.includes(".csv"))) return true;
  if (v.includes("(/") && (v.includes(".md") || v.includes(".csv"))) return true;
  return false;
}

function isPropertyLine(line: string): boolean {
  if (!line || line.trimStart().startsWith("#")) return false;
  const m = PROPERTY_LINE.exec(line);
  if (!m) return false;
  if (m[1]!.includes(":")) return false;
  return true;
}

export function splitNotionPage(text: string): SplitPage {
  const lines = text.replace(/\r\n/g, "\n").split("\n");
  if (lines.length === 0) {
    return {
      h1: "# Untitled",
      h1Text: "Untitled",
      scalarProperties: [],
      bodyLines: [],
    };
  }

  const h1 = lines[0]!;
  let h1Text = h1;
  const hm = /^#\s+(.*)$/.exec(h1);
  if (hm) h1Text = hm[1]!.trim();

  let i = 1;
  while (i < lines.length && !lines[i]!.trim()) i += 1;

  const scalars: [string, string][] = [];
  const relationAndRest: string[] = [];

  if (i < lines.length && isPropertyLine(lines[i]!)) {
    while (i < lines.length) {
      const line = lines[i]!;
      if (!line.trim()) {
        let j = i + 1;
        while (j < lines.length && !lines[j]!.trim()) j += 1;
        if (j < lines.length && isPropertyLine(lines[j]!)) {
          i = j;
          continue;
        }
        break;
      }
      if (!isPropertyLine(line)) break;
      const m = PROPERTY_LINE.exec(line)!;
      const key = m[1]!;
      const val = m[2]!;
      if (isRelationValue(val)) relationAndRest.push(line);
      else scalars.push([key, val]);
      i += 1;
    }
  }

  relationAndRest.push(...lines.slice(i));
  return {
    h1,
    h1Text,
    scalarProperties: scalars,
    bodyLines: relationAndRest,
  };
}
